Release version 2.0

simonask opened the pull request:
Many new features, bugfixes, as well as general refactoring for clarity.

Kijewski committed:
Do not use std::vector in group_adjacent_by()

Using std::vector you have runtime memory allocations, and an O(N) space
complexity.

This PR makes group_adjacent_by() generate ranges, too. You can always
use tranform(to_vector()) if working with vectors is more convenient.

simonask committed:
Added benchmarks

simonask committed:
Many optimizations and features

simonask committed:
CHANGELOG, version number

Kijewski committed:
FIXUP: update description

simonask committed:
Add test for "first after sort"

simonask committed:
Updated CHANGELOG

simonask committed:
Merge branch 'master' into version-2

simonask committed:
CI fix from Kijewski

simonask committed:
Older CMake compatibility

simonask committed:
Make Clang on Linux happy, clang-format

simonask committed:
Merge remote-tracking branch 'Kijewski/pr-O1-space-group_adjacent_by' into version-2

simonask committed:
Update CHANGELOG

simonask commented:
@Kijewski Thanks for your great contributions!

I've added you to a "Contributors" section in the CHANGELOG. Let me know if that is not OK. :-)

Kijewski commented:
@simonask, thank you! :) I really like the idea of your library! It is really well though through and easy to extend.

Kijewski commented on include/rx/ranges.hpp:
```diff
@@ -9,64 +9,99 @@
+
     output_type get() const noexcept; // Get the value at the current position.
+
     void next() noexcept; // Advance to the next position, returning true if there was more.
```
> ~~returning true if there was more.~~

Kijewski commented on include/rx/ranges.hpp:
```diff
@@ -9,64 +9,99 @@
 
     size_t size_hint() const noexcept; // A hint about how many times `next()` can be called.
                                        // It is purely an optimization hint, not the accurate number
                                        // of times.
```
Maybe state explicitly that 0 should be returned if unknown.

Hazurl commented on include/rx/ranges.hpp:
```diff
@@ -9,64 +9,99 @@
+// Unfortunately, this is not compatible with C++20 `[[likely]]` / `[[unlikely]]`, because it has a
+// different syntax structure.
+#if defined(__GNUC__) || defined(__clang__)
+#define RX_LIKELY(cond) __builtin_expect(!!(cond), 1)
```
`__builtin_expect ` returns `long`, so `RX_LIKELY(cond)` hasn't always the same type depending on the compiler.
I would suggest casting the return value to the type of `cond` or casting `cond`  to `long` in the non-buildin case. You could also cast both definitions to `bool`.

simonask replied:
Good point! I will change it to always produce a boolean (although it is probably always going to be used in contexts where explicit conversion to `bool` is happening anyway).

Kijewski commented on include/rx/ranges.hpp:
```diff
@@ -654,107 +831,151 @@ struct transform {
+
         [[nodiscard]] constexpr bool at_end() const noexcept {
-            return i >= n || input_.at_end();
+            return i == n || inner.at_end();
```
Is this construct too ugly? It could aid ranges where calling at_end() is at least a little expensive:

```c++
if (i == n) {
    return true;
} else if (inner.at_end()) {
    i = n;
    return true;
} else {
    return false;
}
```

simonask replied:
Hm yes, but that would require making `i` mutable.

I think many things rely on `at_end()` being cheap already, which is also the rationale for making it `noexcept`.

Hazurl commented on include/rx/ranges.hpp:
```diff
@@ -468,79 +508,203 @@ struct VectorRange {
+        }
+    }
+
+    constexpr output_type get() const noexcept {
```
May that be beneficial to have a version for const& and one for && ?

simonask replied:
Yes, we're already doing that in several places internally, and it would be good to officially support it. :+1:

Hazurl commented on include/rx/ranges.hpp:
```diff
@@ -1354,62 +1620,96 @@ struct foldl {
+    The output type of the input range is required to implement `operator+`.
+*/
 struct sum {
     constexpr sum() noexcept {}
```
It should not be defined or defaulted to keep the constructor trivial.

simonask replied:
:+1:

Hazurl commented on include/rx/ranges.hpp:
```diff
@@ -1354,62 +1620,96 @@ struct foldl {
+struct max : private Compare {
+    template <class C>
+    constexpr max(C&& cmp) noexcept : Compare(std::forward<C>(cmp)) {}
     constexpr max() noexcept {}
```
It should not be defined or defaulted to keep the constructor trivial.

Hazurl reviewed (commented):
README.md#Algorithms should be updated

Kijewski commented on .github/workflows/ccpp.yml:
```diff
@@ -3,17 +3,31 @@ name: C/C++ CI
     - uses: actions/checkout@v1
     - name: install dependencies
-      run: sudo apt-get install libboost-dev doctest-dev
+      run: (sudo apt-get update && sudo apt-get install libboost-date-time-dev doctest-dev ninja-build cmake build-essential g++-9)
```
rx-benchmark needs `libbenchmark-dev` (same below).

simonask replied:
:+1:

Kijewski commented on .github/workflows/ccpp.yml:
```diff
@@ -3,17 +3,31 @@ name: C/C++ CI
       run: ninja -C build
     - name: run tests
-      run: build/rx-ranges-test
+      run: build/rx-ranges-test && build/rx-calendar
```
… && ./rx-benchmark (same below)

simonask committed:
Review feedback: Bool-cast RX_LIKELY/RX_UNLIKELY

simonask committed:
Review feedback: InputRange documentation update

simonask committed:
Review feedback: Default trivial constructors

simonask committed:
Review feedback: Add benchmarks to CI

simonask committed:
.gitignore: Ignore VS litter

simonask committed:
Forward arguments to custom predicates in min/max/sort

simonask committed:
Support non-default-constructible/non-copyable predicates

simonask committed:
Fix compilation on GCC and libstdc++

simonask committed:
CI: Only run "push" action on master branch

simonask committed:
clang-format

simonask committed:
CI: Build/test with Clang/libc++ on Linux

simonask committed:
CI: Don't run benchmarks in Debug mode

simonask committed:
CI: Disable clang-8 benchmark on Linux for now

simonask committed:
CI: Add Clang8 benchmark with libstdc++

Kijewski committed:
Add cycle: create an infinite, looping range

Inspired by [itertools.cycle()][1].

[1]: https://docs.python.org/3.8/library/itertools.html#itertools.cycle

Kijewski committed:
Fix: Make zip accept two same types

It was not possible to use e.g. `zip(seq(), seq())`.

Kijewski committed:
Add: implement zip_longest

Hazurl commented on include/rx/ranges.hpp:
```diff
@@ -1354,62 +1625,116 @@ struct foldl {
+*/
 struct sum {
-    constexpr sum() noexcept {}
+    constexpr sum() = default;
```
I would just remove that line.

Hazurl commented on include/rx/ranges.hpp:
```diff
@@ -1354,62 +1625,116 @@ struct foldl {
+        using type = remove_cvref_t<get_output_type_of_t<R>>;
+        // Note: foldl is reentrant.
+        auto folder = foldl(
+            type{}, [](type accum, auto&& x) constexpr { return accum + x; });
```
`return accum + std::forward<decltype(x)>(x)`

Hazurl commented on include/rx/ranges.hpp:
```diff
@@ -1354,62 +1625,116 @@ struct foldl {
+template <class Compare = std::less<void>>
+struct max : private Compare {
+    template <class... Args>
+    constexpr max(Args&&... args) noexcept : Compare(std::forward<Args>(args)...) {}
```
That overload prevents the copy/move constructor to be called.

simonask replied:
You're right. Maybe it would be simpler to just accept an instance of `Compare` by value. Do you think that would be acceptable?

Hazurl replied:
I think yes

Hazurl replied:
That's also what the standard does in std::set and other

Hazurl commented on include/rx/ranges.hpp:
```diff
@@ -1354,62 +1625,116 @@ struct foldl {
+template <class Compare = std::less<void>>
+struct min : private Compare {
+    template <class... Args>
+    constexpr min(Args&&... args) noexcept : Compare(std::forward<Args>(args)...) {}
```
Same as `max`

Hazurl commented on include/rx/ranges.hpp:
```diff
@@ -1539,40 +1869,54 @@ append(C&)->append<C>;
-    constexpr explicit sort(Compare cmp = Compare{}) noexcept : cmp(std::move(cmp)) {}
+struct sort : private Compare {
+    template <class... Args>
+    constexpr explicit sort(Args&&... args) noexcept : Compare(std::forward<Args>(args)...) {}
```
same as `min` and `max`.

Hazurl commented on include/rx/ranges.hpp:
```diff
@@ -1583,41 +1927,38 @@ sort()->sort<>;
-    constexpr explicit uniq(Compare cmp = Compare{}) noexcept : cmp(std::move(cmp)) {}
+struct uniq : private Compare {
+    template <class C>
+    constexpr explicit uniq(C&& cmp) noexcept : Compare(std::forward<C>(cmp)) {}
```
Still preventing a call to the copy/move constructor

Kijewski committed:
Make append() work for rvalue, too

Kijewski committed:
Make append() work for rvalue, too

Kijewski committed:
Add: empty_range(), a range that is always empty

Kijewski committed:
Add: chain(), return values from multiple ranges

Once the first range is exhausted, values from the second range are
returned, and so on.

Inspired by [itertools.chain()][1].

[1]: https://docs.python.org/3.8/library/itertools.html#itertools.chain

Hazurl committed:
Remove default constructible and copy assignable requirement of idempotent_range

Hazurl committed:
`transform` is not idempotent as `func` is called each time `get` is

Hazurl committed:
Add test about idempotent property of `filter`

Kijewski committed:
Add: flatten(), flatten one level of nesting

Inspired by [more_itertools.flatten()][1].

[1]: https://more-itertools.readthedocs.io/en/stable/api.html#more_itertools.flatten

Kijewski committed:
Add: padded(), chain infinite sequence of constant

Inspired by [more_itertools.padded()][1].

[1]: https://more-itertools.readthedocs.io/en/v7.2.0/api.html#more_itertools.padded

simonask committed:
Merge pull request #11 from Kijewski/pr-add-cycle

 Add cycle: create an infinite, looping range

Kijewski committed:
Examples: use seq in calendar instead of generator

seq() is generic, so why not use it to generate a sequence of dates?

simonask committed:
Merge pull request #20 from Kijewski/pr-seq-in-example

Examples: use seq in calendar instead of generator

simonask committed:
Merge pull request #12 from Kijewski/pr-add-zip_longest

Implement zip_longest, fix zip

simonask committed:
Merge branch 'version-2' into pr-add-padded

simonask committed:
Merge pull request #19 from Kijewski/pr-add-padded

Add: padded(), yield an infinite list of constant values once the input range is exhausted

simonask committed:
Merge pull request #15 from Hazurl/pr-relax-idempotent-range-requirement

Relax `idempotent_range` requirement

simonask committed:
Merge pull request #13 from Kijewski/pr-add-append-rvalues

Make append() work for rvalue, too

simonask committed:
Merge branch 'version-2' into pr-add-chain

simonask committed:
Merge pull request #14 from Kijewski/pr-add-chain

 Add: chain(), return values from multiple ranges

simonask committed:
Introduce advance_by(), and emit ranges from groups

Kijewski committed:
Fix: until() is (most likely) finite

This PR makes until() usable as an input to e.g. to_vector() without
needing a take() guard.

Kijewski committed:
Tests: assert !at_end in all get & next methods

Most ranges will fail catastrophically if get() or next() was called
when they are already at_end(). This PR makes it possible to find
improper usages by adding assertions in all get() and next() methods.

Kijewski committed:
Add: null_sink(), simply discard all elements

simonask committed:
Detect overflows in take::advance_by()

simonask committed:
Re-add missing zero-check in in_groups_of_exactly()

simonask committed:
Fix in_groups_of_exactly

simonask committed:
Update description of in_groups_of

simonask committed:
Simplify skip_n() with advance_by()

simonask committed:
Fix ZipRange::advance_by()

simonask committed:
Minor condition hygiene

simonask committed:
CI: Add MSVC build (#26)

Adds CI jobs to run tests and benchmarks on MSVC/Windows

simonask commented:
Note: It looks like the `rx-calendar` example is currently broken on the `version-2` branch, but that this PR fixes it.

Kijewski committed:
Make sink() usable for single elements

Kijewski committed:
Add get_idempotent_range_type_t for convenience

Kijewski committed:
Add: tee(), copy into container during iteration

simonask committed:
Merge pull request #18 from Kijewski/pr-add-tee

Add: tee(), copy into container during iteration

simonask committed:
Merge pull request #23 from Kijewski/pr-tests-assert-not_at_end

Tests: assert !at_end in all get & next methods

Commit:
Add Hazurl as a contributor

simonask committed:
Merge pull request #27 from simonask/contributor-hazurl

Add Hazurl as a contributor

Commit:
Update docs with examples

simonask committed:
Merge remote-tracking branch 'origin/version-2' into advance_by

simonask committed:
Implement advance_by() for padded, zip_longest

Kijewski committed:
Rename repeat* to fill*

Kijewski committed:
Fix: make fill() and fill_n() work with lvalues

Kijewski committed:
Implify take_n(n, v) to fill(v) | take(n)

Kijewski committed:
Fix: rename member variable end for MSVC

Range-based loops does not work properly in MSVC if the range contains a
member variable `end`. This PR renames such variables into `done`.

simonask committed:
Test ad-hoc lambdas

simonask committed:
Merge pull request #16 from Kijewski/pr-fix-repeat

Fix: make fill() and fill_n() work with lvalues

simonask committed:
Restructure README, add missing docs

simonask committed:
Merge pull request #28 from Kijewski/pr-fix-end

Fix: rename member variable end for MSVC

simonask committed:
Merge branch 'version-2' into advance_by

simonask committed:
Merge remote-tracking branch 'origin/version-2' into advance_by

simonask committed:
advance_by() returns number of successful advances

simonask committed:
Fix uninitialized variable

simonask committed:
Merge pull request #17 from Kijewski/pr-add-flatten

 Add: flatten(), flatten one level of nesting

simonask committed:
Merge pull request #22 from Kijewski/pr-fix-until

 Fix: until() is (most likely) finite

simonask committed:
Merge pull request #24 from Kijewski/pr-add-passthrough

Add: null_sink(), simply discard all elements

simonask committed:
Merge remote-tracking branch 'origin/version-2' into advance_by

simonask committed:
Merge pull request #21 from simonask/advance_by

Introduce advance_by(), and emit ranges from groups

simonask commented:
Note to all: I think it's time to go into "feature freeze" for 2.0.

Before releasing, I want to make the following changes:

- Cleanup of code style, naming conventions, etc.
- Check test coverage.
- Add new benchmarks.
- Implement optimizations (where possible) to try to live up to the "zero-overhead" claim.

@Kijewski @Hazurl You have both been extremely helpful. If you have any outstanding comments or changes, please let me know. :-)

simonask committed:
Merge branch 'version-2' into docs-update

simonask committed:
Add aggregation example

simonask committed:
Sort reference list

simonask committed:
Merge pull request #31 from simonask/docs-update

Update docs for version 2.0

simonask committed:
clang-format

simonask committed:
Clean up forwarding constructors

simonask committed:
Benchmarks: cycle, padded, chain

simonask committed:
Avoid function pointers in ChainRange

simonask committed:
Benchmark any_of, all_of

simonask committed:
Merge pull request #32 from simonask/version-2-benchmarks

Benchmarks, optimizations

simonask committed:
Manually download doctest and google-benchmark

simonask committed:
CI: Don't install doctest/google-benchmark on the system

simonask committed:
Add benchmarks, macOS builds

simonask committed:
Fixup

simonask committed:
Update CHANGELOG

simonask committed:
Merge pull request #33 from simonask/deps-as-external-project

CI: Add benchmarks for Linux+Clang+libcxx, tests and benchmarks for macOS
